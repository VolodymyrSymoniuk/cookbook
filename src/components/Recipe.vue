<template>
  <section class="recipe">
    <div v-if="!recipe" class="recipe-default-text">
      <p>Click on the eye to display the recipe.</p>
      <p>Double-click on the recipe in the list to create a child branch.</p>
      <p>Stop editing recipe description for 3 seconds to save.</p>
    </div>
    <template v-else>
      <div class="recipe-description">
        <NinjaInput
          :key="recipe.id"
          v-model="description"
          class="description-input"
        />
        <span class="material-icons tip-edit">
          create
        </span>
      </div>
      <template v-if="recipe.versions.length">
        <p>Versions:</p>
        <ul>
          <li v-for="(version, index) of recipe.versions" :key="index">
            {{ version }}
          </li>
        </ul>
      </template>
      <small class="recipe-creation-date">
        Created at: <time>{{ recipe.createdAt | date }}</time>
      </small>
    </template>
  </section>
</template>

<script>
import { debounce } from "lodash";
import { mapGetters, mapMutations } from "vuex";
import NinjaInput from "@/common/NinjaInput";

export default {
  name: "Recipe",
  components: { NinjaInput },
  data: () => ({
    description: ""
  }),
  created() {
    this.debouncedUpdateRecipe = debounce(this.updateRecipe, 3000);
  },
  computed: {
    ...mapGetters({ recipe: "currentRecipe" })
  },
  watch: {
    recipe: function({ description }) {
      this.description = description;
    },
    description: function(newDescription, oldDescription) {
      this.debouncedUpdateRecipe(newDescription, oldDescription);
    }
  },
  methods: {
    ...mapMutations(["updateRecipeWithDescription"]),
    updateRecipe(newDescription, oldDescription) {
      if (
        !oldDescription.trim() ||
        !newDescription.trim() ||
        this.recipe.description === newDescription.trim()
      )
        return;

      this.updateRecipeWithDescription({
        description: newDescription,
        id: this.recipe.id
      });
    }
  }
};
</script>
