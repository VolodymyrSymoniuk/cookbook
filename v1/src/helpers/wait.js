export default milliseconds =>
  new Promise(resolve => {
    setTimeout(() => {
      resolve();
    }, milliseconds);
  });
