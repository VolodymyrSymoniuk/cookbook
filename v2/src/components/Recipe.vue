<template>
  <v-card v-if="recipe" min-width="100%" class="mt-4 mx-auto recipe">
    <v-card-title class="headline font-weight-bold">
      <NinjaInput :key="recipe.id" v-model="title" />
    </v-card-title>

    <v-card-text class="font-weight-bold">
      <NinjaInput :key="recipe.id" v-model="description" />
      <v-row class="mr-4 mt-4" align="center" justify="end">
        Created at: {{ recipe.createdAt | date }}
      </v-row>
    </v-card-text>
  </v-card>
  <h2 class="mx-auto mt-4 px-4" v-else>
    Choose the recipe you want to work with, from the cookbook
  </h2>
</template>

<script>
import NinjaInput from "@/common/NinjaInput";
import { mapGetters } from "vuex";

export default {
  name: "Recipe",
  components: { NinjaInput },
  data: () => ({
    title: "",
    description: ""
  }),
  computed: {
    ...mapGetters({ recipe: "activeRecipe" })
  },
  watch: {
    recipe: function(recipe) {
      this.title = recipe.title;
      this.description = recipe.description;
    }
  }
};
</script>
