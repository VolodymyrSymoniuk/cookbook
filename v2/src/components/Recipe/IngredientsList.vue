<template>
  <v-list dense>
    <v-subheader>Ingredients:</v-subheader>
    <v-list-item-group>
      <v-list-item v-for="(ingredient, i) of ingredients" :key="i">
        <v-list-item-content>
          <v-list-item-title>{{ ingredient.name }}</v-list-item-title>
        </v-list-item-content>
        <v-list-item-icon @click.stop="removeIngredient(ingredient.id)">
          <v-icon>mdi-close</v-icon>
        </v-list-item-icon>
      </v-list-item>

      <v-list-item @click="newIngredientFocus">
        <v-list-item-content>
          <input
            ref="newIngredientInput"
            v-model.trim="newIngredient"
            class="disguised-input"
          />
        </v-list-item-content>
        <v-list-item-icon @click.stop="addIngredient">
          <v-icon>mdi-plus</v-icon>
        </v-list-item-icon>
      </v-list-item>
    </v-list-item-group>
  </v-list>
</template>

<script>
import shortid from "shortid";

export default {
  name: "IngredientsList",
  props: ["ingredients"],
  data: () => ({
    newIngredient: "new ingredient"
  }),
  methods: {
    newIngredientFocus() {
      this.$refs.newIngredientInput.focus();
    },
    addIngredient() {
      if (!this.newIngredient) {
        this.newIngredient = "ingredient can not be empty";
        return;
      }

      this.$emit("add", {
        id: shortid.generate(),
        name: this.newIngredient
      });
      this.newIngredient = "new ingredient";
    },
    removeIngredient(id) {
      this.$emit("remove", id);
    }
  }
};
</script>
