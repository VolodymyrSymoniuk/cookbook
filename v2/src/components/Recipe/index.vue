<template>
  <v-card v-if="recipe" min-width="100%" class="my-4 mx-auto recipe">
    <v-card-title class="headline font-weight-bold">
      <NinjaInput :key="recipe.id + version.id" v-model.trim="title" />
    </v-card-title>

    <v-card-text class="font-weight-bold">
      <NinjaInput :key="recipe.id + version.id" v-model.trim="description" />
      <v-row class="mt-4">
        <v-col cols="12" sm="6">
          <IngredientsList
            :ingredients="ingredients"
            @add="addIngredient"
            @remove="removeIngredient"
          />
        </v-col>

        <v-col cols="12" sm="6">
          <v-select
            :key="activeRecipe.versions.length"
            color="teal"
            v-model="version"
            :items="versions"
            item-text="title"
            item-value="id"
            label="Select recipe version"
            return-object
            single-line
          ></v-select>
          <v-btn
            block
            color="white-text teal lighten-1"
            elevation="2"
            @click="updateRecipeData"
          >
            Update recipe
          </v-btn>
          <v-btn
            block
            color="white-text teal lighten-1"
            class="mt-4"
            elevation="2"
            @click="createNewVersion"
          >
            Save as a new version
          </v-btn>
        </v-col>
      </v-row>

      <v-row class="mr-4 mt-4" align="center" justify="end">
        Created at: {{ recipe.createdAt | date }}
      </v-row>
    </v-card-text>
    <Snackbar />
  </v-card>
  <h2 class="mx-auto mt-4 px-4" v-else>
    Choose the recipe you want to work with, from the cookbook
  </h2>
</template>

<script>
import shortid from "shortid";
import { mapGetters, mapActions } from "vuex";
import NinjaInput from "@/common/NinjaInput";
import IngredientsList from "./IngredientsList";
import Snackbar from "@/common/Snackbar";

export default {
  name: "Recipe",
  components: { NinjaInput, IngredientsList, Snackbar },
  data: () => ({
    recipe: null,
    title: "",
    description: "",
    ingredients: [],
    versions: [],
    version: null
  }),
  computed: {
    ...mapGetters(["activeRecipe"])
  },
  watch: {
    activeRecipe: function(activeRecipe) {
      this.recipe = activeRecipe;
      this.versions = [...activeRecipe.versions];
      this.version = activeRecipe.versions[0];
    },
    recipe: function(recipe) {
      this.title = recipe.title;
      this.description = recipe.description;
      this.ingredients = [...recipe.ingredients];
    },
    version: function(version) {
      if (version.id === "default") {
        this.recipe = this.activeRecipe;
      } else {
        // eslint-disable-next-line no-unused-vars
        const { id, ...versionWithoutId } = version;
        this.recipe = { ...this.recipe, ...versionWithoutId };
      }
    }
  },
  methods: {
    ...mapActions({
      showSnack: "snackbar/showSnack",
      updateRecipe: "updateRecipe",
      updateRecipeVersion: "updateRecipeVersion",
      addActiveRecipeVersion: "addActiveRecipeVersion"
    }),
    showNotification(message) {
      this.showSnack({
        text: message,
        color: "success",
        timeout: 2000
      });
    },
    addIngredient(ingredient) {
      this.ingredients.push(ingredient);
    },
    removeIngredient(id) {
      this.ingredients = this.ingredients.filter(i => i.id !== id);
    },
    updateRecipeData() {
      const data = {
        title: this.title,
        description: this.description,
        ingredients: this.ingredients
      };

      if (this.version.id === "default") {
        this.updateRecipe({ data, id: this.recipe.id });
        this.showNotification("Recipe was successfully updated");
      } else {
        this.updateRecipeVersion({
          data,
          recipeId: this.recipe.id,
          versionId: this.version.id
        });
        this.showNotification("Recipe version was successfully updated");
      }
    },
    createNewVersion() {
      const newVersion = {
        id: shortid.generate(),
        title: this.title,
        description: this.description,
        ingredients: this.ingredients,
        createdAt: Date.now()
      };

      this.showNotification("New version was successfully added");

      this.addActiveRecipeVersion(newVersion);
      this.versions.push(newVersion);
      this.version = newVersion;
    }
  }
};
</script>
