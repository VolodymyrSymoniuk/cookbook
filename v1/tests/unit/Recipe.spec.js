import { shallowMount, mount, createLocalVue } from "@vue/test-utils";
import { last } from "lodash";
import Vuex from "vuex";
import Recipe from "@/components/Recipe";
import wait from "../../src/helpers/wait";

const localVue = createLocalVue();
localVue.use(Vuex);
localVue.filter("date", () => "I am formatted date");

let currentRecipe = {
  id: "2WEKaVNO",
  parent: "root",
  description: "description",
  versions: ["version1", "version2", "version3"],
  createdAt: 1610702654321
};

describe("RecipeTree component", () => {
  let mutations, getters, store, wp;

  beforeEach(() => {
    getters = {
      currentRecipe: () => currentRecipe
    };
    mutations = {
      updateRecipeWithDescription: jest.fn(({ description }) => {
        currentRecipe.versions.push(description);
      })
    };
    store = new Vuex.Store({
      getters,
      mutations
    });

    wp = shallowMount(Recipe, { store, localVue });
  });

  it("should render as expected", () => {
    expect(wp).toMatchSnapshot();
  });

  it("should correctly format date by filter", () => {
    expect(wp.html()).toContain("Created at: <time>I am formatted date</time>");
  });

  // it("should correctly update recipe", async () => {
  // const wrapper = mount(Recipe, { store, localVue });
  // const versionsAmount = currentRecipe.versions.length;
  // expect(wrapper.findAll("ul > li").length).toEqual(versionsAmount);
  // const input = wrapper.find(".description-input");
  // input.setValue("new description");
  // input.setValue("newer description");
  // await wait(3100);
  // expect(mutations.updateRecipeWithDescription).toHaveBeenCalledTimes(1);
  // expect(last(currentRecipe.versions)).toEqual("newer description");
  // expect(currentRecipe.versions.length).toEqual(versionsAmount + 1);
  // });
});
