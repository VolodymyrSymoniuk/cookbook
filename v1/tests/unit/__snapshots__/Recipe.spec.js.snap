// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RecipeTree component should render as expected 1`] = `
<section class="recipe">
  <div class="recipe-description">
    <ninjainput-stub value="" class="description-input"></ninjainput-stub> <span class="material-icons tip-edit">
        create
      </span>
  </div>
  <p>Versions:</p>
  <ul>
    <li>
      version1
    </li>
    <li>
      version2
    </li>
    <li>
      version3
    </li>
  </ul> <small class="recipe-creation-date">
    Created at: <time>I am formatted date</time></small>
</section>
`;
